core/modules/loader/BufferUdp.h

```diff
@@ -0,0 +1,191 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ *
+ */
+#ifndef LSST_QSERV_LOADER_BUFFERUDP_H_
+#define LSST_QSERV_LOADER_BUFFERUDP_H_
+
+// system headers
+#include <arpa/inet.h>
+#include <cstring>
+#include <stdexcept>
+#include <memory>
+#include <sstream>
+#include <string>
+
+#include <boost/bind.hpp>
+#include <boost/asio.hpp>
+
+/// The absolute largest UDP message we would send.
+/// Usually, they should be much smaller.
+#define MAX_MSG_SIZE 6000
+
+
+namespace lsst {
+namespace qserv {
+namespace loader {
+
+class MsgElement;
+
+
+/// A buffer for reading and writing. Nothing can be read from the buffer until
+/// something has written to it.
+class BufferUdp {
+public:
+    using Ptr = std::shared_ptr<BufferUdp>;
+
+    BufferUdp() : BufferUdp(MAX_MSG_SIZE) {
+    }
+
+    explicit BufferUdp(size_t length) : _length(length) {
+        _buffer = new char[length];
+        _ourBuffer = true;
+        _setupBuffer();
+    }
+
+    /// If the buffer already contains valid data, advanceWriteCursor()
+    /// must be called with the length of the valid data. Otherwise,
+    /// BufferUdp thinks it has an empty buffer.
+    BufferUdp(char* buf, size_t length) : _buffer(buf), _length(length) {
+        _setupBuffer();
+    }
+
+
+    ~BufferUdp() {
+        if (_ourBuffer) {
+            delete[] _buffer;
+        }
+    }
+
+    void reset() {
+        _setupBuffer();
+    }
+
+
+    /// Return true only if this object owns the buffer.
+    bool releaseOwnership() {
+        if (_ourBuffer) {
+            _ourBuffer = false;
+            return true;
+        }
+        return false;
+    }
+
+
+    void makeOwnerOfBuffer() {
+        _ourBuffer = true;
+    }
+
+
+    /// Return true if there's at least 'len' room left in the buffer.
+    bool isAppendSafe(size_t len) const {
+        return (_wCursor + len) <= _end;
+    }
+
+
+    bool append(const void* in, size_t len) {
+        if (isAppendSafe(len)) {
+            memcpy(_wCursor, in, len);
+            _wCursor += len;
+            return true;
+        }
+        return false;
+    }
+
+
+    void advanceWriteCursor(size_t len) {
+        _wCursor += len;
+        if (not isAppendSafe(0)) {
+            throw new std::overflow_error("BufferUdp advanceWriteCursor beyond buffer len=" +
+                    std::to_string(len));
+        }
+    }
+
+
+    void advanceReadCursor(size_t len) {
+        _rCursor += len;
+        if (_rCursor > _end) {
+            throw new std::overflow_error("BufferUdp advanceReadCursor beyond buffer len=" +
+                                          std::to_string(len));
```

Comment:
do you want to be sure the read cursor does not advance to or beyond the write cursor?

Reply:
If everything has been passed properly to asio, it shouldn't go outside the buffer. But if it has, this lets us know.